### ISSUE TITLE
Provide for omission of resources.

### ISSUE THREAD

**Issue by jmdots** [kubernetes-sigs/kustomize#106]
I'm new to kustomize, so let me know if this even makes sense.

Use case: I'm transforming an all-in-one YAML (`---` delimited).  This works well, except I want to skip one of the resources in the upstream YAML stream.

Something like this:

```
omissionPatches:
- svc.some-public-service.yaml
```

Where `svc.some-public-service.yaml` contains:

```
apiVersion: v1
kind: Service
metadata:
  name: some-public-service
```

In the example scenario, kustomize should not include `svc/some-public-service` in the output stream.

**Comment by Liujingfang1** [comment]
Here is my understanding. By `omissionPatches`, you want kustomize to delete certain resources declared in the base.
This is a good feature to have.

As a workaround, you can remove the declaration for `some-public-service` from the all-in-one YAML.

**Comment by Liujingfang1** [comment]
I'd also like to you how you get the all-in-one YAML. You can split it and group the resources into different files. 
Then create kustomization based on the new files.

**Comment by Liujingfang1** [comment]
https://github.com/kubernetes-sigs/kustomize/blob/master/docs/eschewedFeatures.md

**Comment by monopole** [comment]
@jmdot - i just think this is hard to do in the general case, and want to think carefully and get more experience before we try it.

You're workaround - fork the upstream - is exactly the right thing to do.

**Comment by danp** [comment]
Apologies for commenting on a closed issue, please let me know if I should open a new one. I think I ran into what this issue was about by trying something like this:

```
apiVersion: v1
kind: Service
metadata:
  name: some-public-service
$patch: delete
```

It works sometimes, but not others, maybe depending on whether the resource is a CRD. Would supporting that be useful? Or maybe it only works sometimes because I'm misunderstanding the patch interaction somehow.

**Comment by KarstenSiemer** [comment]
This indeed works, but the resulting yaml bundle is not usable by kubernetes anymore.
This is what the yaml looks like in the bundle now:
```
---
{}
```
applying this to kubernetes results in:
`error: error validating "STDIN": error validating data: [apiVersion not set, kind not set]; if you choose to ignore these errors, turn validation off with --validate=false`
turning off validation is not a feasible solution. Also i'd like to be able to directly call kustomize without any scripts in between, because i am using argocd for rollout

**Comment by anarcher** [comment]
I think that providing for omission of resources is necessary to avoid some kind of workaround. (e.g. make many directions)

Does kustomize have a plan to support this?

**Comment by gobengo** [comment]
Being able to patch a resource with a top-level `$patch: delete` would be really useful and sensible way of allowing this!

**Comment by anarcher** [comment]
IMHO, using annotation of omission can be useful. 

```
kind: Deployment
metadata:
     name: ingress-nginx
     annotations:
          kustomize.config.k8s.io/overlay-behavior:  delete

```

**Comment by dobesv** [comment]
In https://github.com/kubernetes-sigs/kustomize/blob/master/docs/eschewedFeatures.md you can see that kustomize has specifically listed this as something it will not support.

**Comment by ls-brentsmith** [comment]
TL;DR - Create a minimal base and add overlays to add on additional layers.  Optionally, use multiple bases and specify the ones you want to use in your overlays. 

---
You'd need to break up your manifests, but Kustomize is very good at remerging them.
Then you can have a base that uses the minimum required and add on other resources in overlays.

Create base and overlay directories
```
:➜  cd $(mktemp -d)
:➜  mkdir -p base overlays/add_service
```

Create base and a base resource
```
:➜  cat <<EOF > base/kustomization.yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
- deployment.yaml
EOF

:➜  cat <<EOF > base/deployment.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: some-deployment
EOF
```
Create an overlay and overlay service
```
:➜  cat <<EOF > overlays/add_service/svc.some-public-service.yaml
apiVersion: v1
kind: Service
metadata:
  name: some-public-service
EOF

:➜ cat <<EOF > overlays/add_service/kustomization.yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
bases:
- ../../base
resources:
- svc.some-public-service.yaml
EOF
```

Using kustomize on base would not add the service.
```
:➜  kubectl kustomize base
apiVersion: apps/v1
kind: Deployment
metadata:
  name: some-deployment
```

Using kustomize on the add_service would
```
:➜ kubectl kustomize overlays/add_service/
apiVersion: v1
kind: Service
metadata:
  name: some-public-service
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: some-deployment
```

You can also do stuff like have multiple bases in the same repo and have overlays refer to some bases, do patches, add other resources, etc.

Hopefully this illustrates why they chose specifically not to support deleting a resource from base, as the readability would suffer a lot if you could add, delete, and patch in an overlay, while also including multiple bases.

**Comment by kferrone** [comment]
```yaml
apiVersion: v1
kind: Service
metadata:
  name: some-public-service
$patch: delete
```
This works very well. I wasn't even able to reproduce any of the issues above.

**Comment by amorozkin** [comment]
May be it will be usefull for someone in some cases )

`kustomize build overlays/dev/ | yq 'select(.kind != "CustomResourceDefinition")'`
